### PR TITLE
[WIP] Data Builders Work, Recipes, Conditions, TODO: BlockStates/Models.

### PR DESCRIPTION
**Pull request by LexManos**
This is my work to cleanup the Recipe system, as well as expose some data providers in useful ways to modders.

Recipes:

Conditions have been revamped to have a full serializer class. All the conditions added by Forge are now stand alone classes instead of lambdas. As well as a IConditionBuilder that contains a bunch of helper macros for building conditions. That part is a WIP but I kinda like how it turned out {see the DataGeneratorTest class} 

Ingredients didn't change much, except for moving from fields/lambdas in CraftingHelper to their own Classes.

ConditionalRecipe is a new recipe type I added that will use the first recipe that passes all conditions. This is the preferred way to add conditions to recipes just because it doesn't change the json format of other recipes. It just nests them. This also allows you to have multiple variants of the same recipe name. Configured by conditions. Only one [or none] will be used. But it should help linking advancements/criterias/whatever as the name will be stable.

This also has a change to Ingredient to use a fake Barrior item for empty tags instead of error. I'll probably pull this out into it's own commit. It just makes some data generation easier as the tag doesn't need to actually have anything in it now to serialize properly.

BlockState.json, and Model jsons need data generators. There are two parts the Serializer implementation, and the builders. If someone wants to step up and help create those that would be great. 

Opened as a PR to get discussion and ideas flowing. As well as to let people talk while i'm out of town for the next week.

### CODE REVIEW

File: src/main/java/net/minecraftforge/common/ForgeMod.java [hunk 1]
```diff
@@ -172,4 +187,22 @@ public void gatherData(GatherDataEvent event)
...
+    {
+        CraftingHelper.register(AndCondition.Serializer.INSTANCE);
+        CraftingHelper.register(FalseCondition.Serializer.INSTANCE);
+        CraftingHelper.register(ItemExistsCondition.Serializer.INSTANCE);
+        CraftingHelper.register(ModLoadedCondition.Serializer.INSTANCE);
+        CraftingHelper.register(NotCondition.Serializer.INSTANCE);
+        CraftingHelper.register(OrCondition.Serializer.INSTANCE);
+        CraftingHelper.register(TrueCondition.Serializer.INSTANCE);
+
+        CraftingHelper.register(new ResourceLocation("forge", "compound"), IngredientNBT.Serializer.INSTANCE);
```
Comment (TeamSpen210) [inline src/main/java/net/minecraftforge/common/ForgeMod.java]: These two look back to front. Is this correct?

Comment (LexManos) [reply src/main/java/net/minecraftforge/common/ForgeMod.java]: Yes.. yes they are.. Would of caught in my next round of tests but good eye.

**Comment by EpicSquid** [comment]
How will recipes handle tags with the fake item? Will there be a default conditional available in forge to handle isTagEmpty or isTagAvailable (or is there a better way to handle that)

**Comment by LexManos** [comment]
We could add a tagEmpty conditional. But honestly tags should never be empty if used in a recipe. The recipe creators have access to their environment. They should know whats up.
What cases would you have for checking if a tag is empty rather then just adding something to the tag?

**Comment by KnightMiner** [comment]
One reason you may want to skip a recipe if a tag is empty is for compatibility with a common material. Let's say I have a block crafted with iron, but I want to also support crafting it with copper if another mod adds it.

My example mod does not need copper outside of that one recipe, so I have the copper recipe only load if the copper tag is not empty. You could of course make a common tag with both copper and iron, but that requires that they are both used in the same ratio (so I cannot make it cost twice as much copper)

Additionally, you could use an empty check to have one recipe if copper is present (the tag is not empty), then load a fallback using iron ingots if the tag is empty.

**Comment by EpicSquid** [comment]
The only example of an empty tag I have seen (in another mod that caused me issues) was a mod adding a block made from copper but without adding copper itself. Having a "is tag empty" or "is tag available" would help for mods trying to do general integration like that.

(I'm open to better solutions if there are any btw)

**Comment by LexManos** [comment]
Fine, added a forge:tag_empty condition. {Negatable with forge:not}
I was hoping to push this to master at the same time as Fluids, but if that doesn't get done this weekend i'll push this separate.

**Comment by LexManos** [comment]
Expanded the conditional system to the Advancement loading, as recipes and advancements are tightly tied, it makes sense to have it there as well.

**Comment by LexManos** [comment]
This PR, and it's sister [Fluid Rewrite](https://github.com/MinecraftForge/MinecraftForge/pull/5983) are both good to go IMO. And can be merged, I just want them merged at the same time/quickly together. As they are both major binary breaking changes.

Once these two are in, we'll do a sweep of basic bugs/prs and call it good for a RB.

**Comment by SilverDavidMC** [comment]
How will this change affect existing mods that don't use the data generators?

**Comment by TeamSpen210** [comment]
It doesn't at all, they're entirely optional. You need to listen to the events to have anything happen.

**Comment by SilverDavidMC** [comment]
Alright thanks. The line that said "BlockState.json, and Model jsons need data generators." had me a bit confused.

**Comment by LexManos** [comment]
It will break any mod that adds custom recipe conditions. As the API for that has changed/been cleaned up. Namely, this class has changed: https://github.com/MinecraftForge/MinecraftForge/pull/6014/files#diff-2b5ece9fa863e3ecacffd627afe1c3d9

But, besides that its pretty compatible and just an extension of what already existed.
The line about blockstate and model jsons needing generators, is because vanilla doesn't provide any. And it would be nice to have data generators where you can do something like 
`new BlockStateBuilder(myBlock).addState(myBlock.withProperty(color, red), new ModelBuilder().parent("cube/all").texture("myblock/red")).build();` and have it generate the necessary jsons, validate that all blockstates have a model, and that the textures exists.
There are a lot of cool things that can be done with data generators. They just need to be written. BUT as that extra stuff, and adding doesn't break binary compat, we can add them in later. {If someone wants to start a PR/work on this feel more then free}

**Comment by gigaherz** [comment]
Merged in e784a7f60

**Pull request closed without merging**